### DISCUSSION TITLE
FV vs FE performance

### DISCUSSION THREAD

**Discussion by heinono1** [idaholab/moose#25235] [Q&A General] [answered]
Hi. It seems like my FV version of the ferret app is running well and producing output that seems OK. So I am running checks on identical systems to compare execution speed. I was surprised to see that for a small-ish mesh (200 k nodes), the FE version was ~ 3 x faster than the FV version(!). When I make the mesh larger (just doubling the points in all three directions) to 1.6 M nodes, the FE version is still faster, but now by not much. This seems to imply there is a cross-over for sufficiently large meshes at some few million or so nodes. Does this seem OK or does it seem to jive with others' experience with FV?

**Comment by GiudGiud** [comment, marked as answer]
@lindsayad @pbalest 
For comments on NS experience 

If there is something special to Ferret at play here we ll want to look at the perfgraph output for these cases

**Reply by GiudGiud** [reply]
@snschune as well for comments on FE vs FV

**Reply by heinono1** [reply]
I am attaching performance graph outputs for the two runs. In summary, the main difference in performance seems to be in NonlinearSystemBase::Kernels where the FV version consumed 2862.218 s while the FE version consumed only 317.395 s, a difference of about a factor of nine. On the other hand, for FEProblem::solve the FV version used 501.765 s while the FE version needed 1142.818, so here the FV version was much faster.  So it seems I need to try to optimize the FV kernels better.

Here are the data for the FE version:
```
Heaviest Branch:
--------------------------------------------------------------------------------------------------------------------------------------------------------------------
|                             Section                            | Calls |   Self(s)  |   Avg(s)   |    %   | Mem(MB) |  Total(s)  |   Avg(s)   |    %   | Mem(MB) |
--------------------------------------------------------------------------------------------------------------------------------------------------------------------
| FerretApp (main)                                               |     1 |      0.011 |      0.011 |   0.00 |       3 |   2526.443 |   2526.443 | 100.00 |    2446 |
|   MooseApp::run                                                |     1 |      0.000 |      0.000 |   0.00 |       0 |   2526.430 |   2526.430 | 100.00 |    2443 |
|     MooseApp::execute                                          |     1 |      0.000 |      0.000 |   0.00 |       0 |   2522.145 |   2522.145 |  99.83 |    2203 |
|       MooseApp::executeExecutioner                             |     1 |      0.083 |      0.083 |   0.00 |       0 |   2522.145 |   2522.145 |  99.83 |    2203 |
|         Transient::PicardSolve                                 |    89 |      6.432 |      0.072 |   0.25 |       0 |   2490.261 |     27.980 |  98.57 |      26 |
|           FEProblem::solve                                     |    89 |   1142.818 |     12.841 |  45.23 |    -171 |   2340.807 |     26.301 |  92.65 |      26 |
|             FEProblem::computeJacobianInternal                 |   276 |      0.008 |      0.000 |   0.00 |       0 |    837.776 |      3.035 |  33.16 |     197 |
|               FEProblem::computeJacobianTags                   |   276 |      6.291 |      0.023 |   0.25 |       0 |    837.768 |      3.035 |  33.16 |     197 |
|                 NonlinearSystemBase::computeJacobianTags       |   276 |      0.006 |      0.000 |   0.00 |       0 |    831.468 |      3.013 |  32.91 |     197 |
|                   NonlinearSystemBase::computeJacobianInternal |   276 |    452.490 |      1.639 |  17.91 |     197 |    831.462 |      3.013 |  32.91 |     197 |
|                     FEProblem::checkExceptionAndStopSolve      |   828 |    378.971 |      0.458 |  15.00 |       0 |    378.971 |      0.458 |  15.00 |       0 |
--------------------------------------------------------------------------------------------------------------------------------------------------------------------

Heaviest Sections:
-----------------------------------------------------------------------------------------------------
|                    Section                   | Calls |   Self(s)  |    Avg.    |    %   | Mem(MB) |
-----------------------------------------------------------------------------------------------------
| FEProblem::solve                             |    89 |   1142.818 |     12.841 |  45.23 |    -171 |
| NonlinearSystemBase::computeJacobianInternal |   277 |    453.955 |      1.639 |  17.97 |     197 |
| FEProblem::checkExceptionAndStopSolve        |  3643 |    403.867 |      0.111 |  15.99 |       0 |
| NonlinearSystemBase::Kernels                 |   454 |    317.395 |      0.699 |  12.56 |       0 |
| FEProblem::computeUserObjects                |  2181 |    137.958 |      0.063 |   5.46 |       0 |
| Console::outputStep                          |   554 |     15.466 |      0.028 |   0.61 |       0 |
| FEProblem::computeResidualTags               |   454 |      9.010 |      0.020 |   0.36 |       0 |
```
Here are the heaviest branch and heaviest sections for the FV version:
Heaviest Branch:
```
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------
|                               Section                               | Calls |   Self(s)  |   Avg(s)   |    %   | Mem(MB) |  Total(s)  |   Avg(s)   |    %   | Mem(MB) |
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------
| FerretApp (main)                                                    |     1 |      0.015 |      0.015 |   0.00 |       4 |   3843.120 |   3843.120 | 100.00 |    1129 |
|   MooseApp::run                                                     |     1 |      0.000 |      0.000 |   0.00 |       0 |   3843.105 |   3843.105 | 100.00 |    1125 |
|     MooseApp::execute                                               |     1 |      0.000 |      0.000 |   0.00 |       0 |   3832.625 |   3832.625 |  99.73 |     462 |
|       MooseApp::executeExecutioner                                  |     1 |      0.081 |      0.081 |   0.00 |       0 |   3832.625 |   3832.625 |  99.73 |     462 |
|         Transient::PicardSolve                                      |    89 |      0.402 |      0.005 |   0.01 |       0 |   3782.768 |     42.503 |  98.43 |     418 |
|           FEProblem::solve                                          |    89 |    501.765 |      5.638 |  13.06 |    8321 |   3591.998 |     40.360 |  93.47 |     418 |
|             NonlinearSystemBase::computeResidualAndJacobianInternal |   358 |      0.057 |      0.000 |   0.00 |       0 |   2613.061 |      7.299 |  67.99 |     388 |
|               NonlinearSystemBase::Kernels                          |   358 |   2408.305 |      6.727 |  62.67 |     422 |   2408.305 |      6.727 |  62.67 |     422 |
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------

Heaviest Sections:
----------------------------------------------------------------------------------------------
|                Section                | Calls |   Self(s)  |    Avg.    |    %   | Mem(MB) |
----------------------------------------------------------------------------------------------
| NonlinearSystemBase::Kernels          |   447 |   2862.218 |      6.403 |  74.48 |     688 |
| FEProblem::solve                      |    89 |    501.765 |      5.638 |  13.06 |    8321 |
| FEProblem::checkExceptionAndStopSolve |  1430 |    230.782 |      0.161 |   6.01 |       0 |
| FEProblem::outputStep                 |   358 |    135.768 |      0.379 |   3.53 |       0 |
| FEProblem::computeUserObjects         |  1615 |     62.802 |      0.039 |   1.63 |       0 |
| Console::outputStep                   |   628 |     10.583 |      0.017 |   0.28 |       0 |
| NonlinearSystemBase::residualSetup    |   447 |      7.236 |      0.016 |   0.19 |   -8591 |
----------------------------------------------------------------------------------------------
```

**Reply by heinono1** [reply]
@GiudGiud , @lindsayad, @pbalest, @snschune : it looks like the poorer performance for the FV version comes from worse performance of the preconditioner, so many more linear iterations per nonlinear iteration for the FV system vs. FE system. For the FE system, the bjacobi preconditioner works really well. For FV I am using asm (as suggested by @lindsayad). Tweaking a little does not seem to improve performance; boomeramg does not work well. BTW, I did complete checking Jacobians, and they are OK :-)

**Reply by heinono1** [reply]
@GiudGiud , @lindsayad, @pbalest, @snschune : correction. The asm preconditioner for the FV system does fewer linear iterations than the bjacobi preconditioner for the FE system (by a significant amount). What takes more time in the FV system is the nonlinear solve. I am stumped as to how to improve that.

**Reply by GiudGiud** [reply]
Forming the Jacobian could be the slow step.
You could try a few things:
- computing the Jacobian and the residual together
https://mooseframework.inl.gov/moose/source/systems/NonlinearSystem.html#57f02cde-aa77-4218-a6ac-caf8850d27ca
- re-using the preconditioner inside a solve
https://mooseframework.inl.gov/moose/source/systems/NonlinearSystem.html#b9b3ad4e-c468-4e57-8f4f-328f2b7326f3

How many variables do you have again? And you are solving them fully coupled right?
You have both elemental and flux fv kernels right?
and in your kernels, you have a bunch of gradient evaluations?

**Reply by heinono1** [reply]
Thanks, Guillaume.
I am computing the Jacobian and residual together as in the link. I'll try re-using the preconditioner.
I have four variables (three for the components of the magnetization director, and one for the scalar magnetostatic potential). There is one flux kernel that contains the magnetization directors and their gradients, two volume kernels with magnetization directors and (one of the two) with gradients of the potential, and then there is a diffusion flux kernel for the potential that also has the magnetization directors as source term. I am solving them fully coupled.

**Reply by GiudGiud** [reply]
I think we have a paramter on the FV variables to cache gradients. You can try to see if that improves things. 
If we want to stop shooting in the dark we ll need to profile
https://mooseframework.inl.gov/application_development/profiling.html

**Reply by lindsayad** [reply]
cell gradients are cached by default

**Reply by lindsayad** [reply]
@heinono1 
- do you take *more* nonlinear iterations with FV or is it that each nonlinear iteration takes longer (or both)?
- Are your finite element Jacobians hand-coded? 
- How large are your finite volume cell stencils? Do you just access data from one neighbor layer away?

**Reply by heinono1** [reply]
- The FE and FV systems take the same number of nonlinear iterations almost all the time (typically 2 or 3)
- The FE Jacobians are hand-coded.
- The FV stencils are default (eg one ghost layer).

Re-using preconditioner did not make any difference. Maybe slowed things down by epsilon. I am going to try hand-coding a Jacobian in one kernel to see if that makes a difference. 

I can't do the profiling now. I started building new MOOSE but my company in its wisdom had blocked bitbucket so I can't pull libmesh. I have opened a ticket with IT and I should have bitbucket accessible early next week.

**Reply by snschune** [reply]
From my (admittedly too far removed) point of view, the main difference between FEM and FV is that FV has a threaded internal face loop in addition to an element loop. I would target profiling on how expensive the face loop is compared to the elemental loop. 
If this is true, the solution would not be on the app level but on the framework level.

**Reply by lindsayad** [reply]
> I can't do the profiling now. I started building new MOOSE but my company in its wisdom had blocked bitbucket so I can't pull libmesh. I have opened a ticket with IT and I should have bitbucket accessible early next week.

libMesh and PETSc are on github.com and gitlab.com respectively. However, I do believe our configuration of PETSc does pull down some packages from bitbucket so perhaps that is the issue you're running into?

**Reply by lindsayad** [reply]
> I am going to try hand-coding a Jacobian in one kernel to see if that makes a difference.

Unfortunately I do not think we even provide the ability for users to supply a hand-coded Jacobian for FV

**Reply by heinono1** [reply]
Yeah that seemed to be the case....I am trying a different tack in which I use a an elemental (volume) approximation instead of flux kernels. But that means I need access to shape functions and their gradients in an FVElementalKernel. I have not figured out how to do that yet.

**Reply by heinono1** [reply]
@lindsayad yeah I saw that there are repos on gituhub and gitlab. It was indeed the bitbucket packages for PETSc that stopped me. Rather than trying to redirect the PETSc build I'm just going to wait for IT security to release bitbucket as a "safe" site.

**Reply by lindsayad** [reply]
Thinking about this, if you override computeJacobian in your objects, then you can do whatever you want to compute the Jacobian. The biggest hindrance I see is that the variable gradient computation is hard coded to be AD

**Reply by heinono1** [reply]
Yeah I think I am not going to go down that path but first of all focus on comparing the FV code as I have it with the FE version for some large production run to see how that goes.

**Reply by GiudGiud** [reply]
> .I am trying a different tack in which I use a an elemental (volume) approximation instead of flux kernels.

this should be possible. Once on an element you know what neighbors you have and you can use the functor arguments to get evaluations there.
Do you have a diffusion term? Or is it advection?

**Reply by heinono1** [reply]
Just a diffusion term (Poisson equation).

**Reply by GiudGiud** [reply]
We could compute a volumetric laplacian term using the Functor system. I imagine the stencil is going to end up being fairly similar to what you have with flux kernels. 
If you're careful it can still be conservative? (eg effective flux into current cell from another - however that s defined in this volumetric idea - is the same as the flux from the other cell into the current cell)

I m not sure it s worth exploring tbh

**Reply by heinono1** [reply]
Thanks, Guillaume. I think I will not dive into this line of work. I will rebuild moose and ferret with profiling to try to sort out where exactly execution spends time. I do wonder, though, is something like this has been seen before (poor or lesser performance of FV than FE). If not, it seems more likely that the problems are with my kernels, and they need to be fixed, than anything else.

**Reply by lindsayad** [reply]
I think that for the same number of degrees of freedom it would not be surprising to see the finite element calculations run faster. With compact stencils the finite element calculations can easily cache and re-use calculations. Finite volume, with a more arbitrarily sized and sometimes large stencil, uses a lot of on-the-fly calculations to make the code maintainable and clear. There could be some more reimagining of the finite volume assembly process in the next fiscal year, especially if we continue to pursue SIMPLE solvers, e.g. @grmnptr's https://github.com/idaholab/moose/pull/22699

**Reply by heinono1** [reply]
Thanks for all your comments! Much appreciated!
